src/libsyntax/parse/parser/item.rs

```diff
@@ -1520,14 +1522,15 @@ impl<'a> Parser<'a> {
         if self.eat(&token::OpenDelim(token::Brace)) {
             while self.token != token::CloseDelim(token::Brace) {
                 let field = self.parse_struct_decl_field().map_err(|e| {
-                    self.recover_stmt();
+                    self.consume_block(token::Brace, false);
```

Comment:
Btw... this function and friends are a mess; I tried to convert this to use the list parsing utilities before but the success was not complete; I'll probably give it a go some time in the future.